cache_cleaner/cache_cleaner.py

```diff
@@ -17,68 +17,111 @@
                             printing which files would be deleted without actually deleting.
 
 """
-
 import os
-import re
 import time
 
 import daiquiri
 import docopt
+import humanize
 
-import simulfs
-
-
-daiquiri.setup(level=os.environ.get('LOG_LEVEL', 'INFO'))
-logger = daiquiri.getLogger(__name__)
-
+from utils import parse_max_cache_size_arg, delete_dir_if_empty, delete_file
 
-def parse_max_cache_size_arg(value):
-    m = re.match(r'([0-9]+)(K|M|G|T)', value)
-    if m is not None:
-        size = int(m.group(1))
-        suffix = m.group(2)
-        if suffix == 'K':
-            return size * 1024
-        elif suffix == 'M':
-            return size * 1024 * 1024
-        elif suffix == 'G':
-            return size * 1024 * 1024 * 1024
-        elif suffix == 'T':
-            return size * 1024 * 1024 * 1024 * 1024
-    return int(value)
+daiquiri.setup(level=os.environ.get("LOG_LEVEL", "INFO"))
+logger = daiquiri.getLogger("cache_cleaner")
 
 
 def main():
     args = docopt.docopt(__doc__)
 
     now = time.time()
-    max_age = int(args['--max-age']) * 24 * 60 * 60
-    cache_path = args['--path']
-    max_cache_size = parse_max_cache_size_arg(args['--max-size'])
+    max_age = int(args["--max-age"]) * 24 * 60 * 60
+    cache_path = args["--path"]
+    max_cache_size = parse_max_cache_size_arg(args["--max-size"])
 
-    force = bool(args['--force'])
+    force = bool(args["--force"])
     if force:
-        os.environ['X-RUN-CACHE-CLEANER'] = 'True'
+        os.environ["X-RUN-CACHE-CLEANER"] = "True"
+
+    total_size = 0
+    initial_pass = True
+    while initial_pass or total_size > max_cache_size:
```

Comment:
There's a lot going on in this `while` loop, and I think it'd be easier to read if we broke it up into a few distinct functions. And the `total_size` variable is a bit fiddly to follow.

How about something like:

```python
while get_total_size(path) > max_size:
    delete_all_files(in=path, less_than=max_age)

    current_size = get_total_size(path)
    if current_size > max_size:
        max_age = get_new_max_age(old_max_age=max_age, target_size=max_size, current_size= current_size)
```

Reply:
This sort of a thing, while very readable, would require that `get_total_size` and `delete_all_files` close over variables that they mutate (`total_size`, `initial_pass`) - I think that is liable to cause more issues than if the loop contains more stuff and all functions are pure.

The `total_size` stuff is definitely tricky to follow but using it like this is an optimisation which is sort of necessary - we build it up on the first pass and then decrease it after that, and the `return` stuff is needed to short circuit (and delete no more than is required). I considered putting the mutable variables in a dict so that functions could mutate them but that seemed even more unclear.

I've moved the `new_max_age` logic into a function and reordered the loop contents so that the branching isn't as deep and there isn't a `continue`, which I think was confusing. What do you think?

Reply:
It’s better, but I’m still not in love with it.

Part of my feeling was that we might do away with `total_size` and `initial_pass` – defer to the OS to get the size. We don’t handle decrementing or tracking `total_size` ourselves; we just call a shell utility to get the value.

Reply:
I looked into that but it would double+ the number of inode stats (as we'd still be looking for the `atime`), which are the main bottleneck here. There isn't any filesystem magic that the OS can do which we can't; a `du -sh` is super slow (the same speed as us, but without doing the other stuff we need to do while we loop) because of all those stats.

Another option would be to use `shutil.disk_usage`, which looks at the mounted filesystem space via `statvfs`. However, I don't think it's appropriate for us as we're looking at a path's contents, which doesn't necessarily coincide with a mount point. That's a blocker, but even disregarding that it would be unwise to call `statvfs` every file loop iteration, and we're still `stat`-ing for the `atime`, so there's a perf penalty.

This is an optimised loop, because this is a rare case where optimisation is really required, and unfortunately that does almost always come at the expense of clarity. Not sure what else I can do here without either hiding complexity in weird ways or fairly significantly hurting the perf...

Reply:
Hmm.

Yeah, and I guess you end up duplicating loop logic. Blergh.

I still don't love this code, but I'm convinced we can't make it better without a lot more work.